File: frontend/src/api/planning.ts
import { getJson, postJson } from './http-client';

export interface Store {
  readonly store_id: string;
  readonly store_code: string;
  readonly store_name: string;
  readonly city: string;
  readonly region: string;
  readonly country: string;
  readonly timezone: string;
}

export interface Product {
  readonly product_id: string;
  readonly sku: string;
  readonly product_name: string;
  readonly category: string;
  readonly subcategory: string;
  readonly brand: string;
  readonly unit_cost: string;
  readonly unit_price: string;
}

export interface DemandPoint {
  readonly period_start: string;
  readonly demand_units: number;
}

export interface DemandSeries {
  readonly store_id: string;
  readonly product: Product;
  readonly policy: {
    readonly lead_time_days: number;
    readonly safety_stock: number;
    readonly reorder_point: number;
    readonly minimum_order_quantity: number;
    readonly order_multiple: number;
    readonly service_level_target: number;
  };
  readonly observations: readonly DemandPoint[];
  readonly summary: {
    readonly period_start: string;
    readonly period_end: string;
    readonly total_units: number;
    readonly average_monthly_units: number;
    readonly minimum_monthly_units: number;
    readonly maximum_monthly_units: number;
    readonly zero_demand_months: number;
  };
}

export interface ForecastMetrics {
  readonly wape: number | null;
  readonly mase: number | null;
  readonly rmse: number;
  readonly bias: number | null;
  readonly interval_coverage: number | null;
  readonly validation_points: number;
}

export interface ModelEvaluation {
  readonly model_name: string;
  readonly status: 'succeeded' | 'failed';
  readonly validation_origins: number;
  readonly metrics: ForecastMetrics | null;
  readonly failure_detail: string | null;
}

export interface ForecastPoint {
  readonly period_start: string;
  readonly forecast_value: number;
  readonly lower_bound: number;
  readonly upper_bound: number;
  readonly horizon: number;
}

export interface ForecastPreview {
  readonly forecast_id: string;
  readonly store_id: string;
  readonly product: Product;
  readonly selected_model: string;
  readonly training_cutoff: string;
  readonly horizon_months: number;
  readonly interval_level: number;
  readonly intermittent_demand: boolean;
  readonly validation_origins: number;
  readonly selected_metrics: ForecastMetrics;
  readonly evaluations: readonly ModelEvaluation[];
  readonly forecasts: readonly ForecastPoint[];
}

export interface ForecastPreviewRequest {
  readonly store_id: string;
  readonly product_id: string;
  readonly horizon_months: number;
  readonly interval_level: number;
}

export interface DemandSignal {
  readonly signal_id: string;
  readonly store_id: string;
  readonly product_id: string;
  readonly sku: string;
  readonly signal_type: 'trend' | 'volatility' | 'seasonality' | 'anomaly';
  readonly direction:
    'growing' | 'declining' | 'stable' | 'volatile' | 'seasonal' | 'spike' | 'drop';
  readonly severity: 'info' | 'watch' | 'warning';
  readonly period_start: string | null;
  readonly title: string;
  readonly description: string;
  readonly metric_name: string;
  readonly metric_value: number;
  readonly baseline_value: number | null;
  readonly threshold_value: number;
  readonly evidence: Readonly<Record<string, string | number>>;
}

export interface InventoryPosition {
  readonly store_id: string;
  readonly product_id: string;
  readonly sku: string;
  readonly snapshot_at: string;
  readonly on_hand: number;
  readonly allocated: number;
  readonly on_order: number;
  readonly on_order_due_within_lead_time: number;
  readonly next_expected_receipt_at: string | null;
  readonly lead_time_days: number;
  readonly safety_stock: number;
}

export interface InventoryRisk {
  readonly risk_id: string;
  readonly store_id: string;
  readonly product_id: string;
  readonly sku: string;
  readonly classification: 'stockout' | 'below_safety_stock' | 'healthy' | 'excess';
  readonly severity: 'info' | 'watch' | 'warning';
  readonly snapshot_at: string;
  readonly replenishment_arrival_at: string;
  readonly available_inventory: number;
  readonly on_order_due_within_lead_time: number;
  readonly forecast_demand_during_lead_time: number;
  readonly projected_inventory: number;
  readonly safety_stock: number;
  readonly coverage_months: number | null;
  readonly excess_coverage_threshold_months: number;
  readonly selected_model: string;
  readonly forecast_id: string;
  readonly training_cutoff: string;
  readonly forecast_horizon_months: number;
}

export interface PlanningException {
  readonly exception_id: string;
  readonly store_id: string;
  readonly product_id: string;
  readonly sku: string;
  readonly exception_type:
    | 'potential_stockout'
    | 'below_safety_stock'
    | 'excess_inventory'
    | 'demand_spike'
    | 'demand_decline'
    | 'forecast_uncertainty'
    | 'forecast_bias';
  readonly severity: 'medium' | 'high' | 'critical';
  readonly priority_score: number;
  readonly status: 'open' | 'acknowledged' | 'resolved' | 'dismissed';
  readonly relevant_period: string;
  readonly title: string;
  readonly description: string;
  readonly metric_name: string;
  readonly metric_value: number;
  readonly threshold_value: number;
  readonly evidence: Readonly<Record<string, string | number>>;
  readonly related_signal_id: string | null;
  readonly related_risk_id: string | null;
  readonly related_forecast_id: string | null;
  readonly created_at: string;
}

export function listStores(signal?: AbortSignal): Promise<Store[]> {
  return getJson<Store[]>('/api/v1/stores', signal);
}

export function searchProducts(
  storeId: string,
  query: string,
  signal?: AbortSignal,
): Promise<Product[]> {
  const parameters = new URLSearchParams({ store_id: storeId, query, limit: '50' });
  return getJson<Product[]>(`/api/v1/products?${parameters.toString()}`, signal);
}

export function getDemandSeries(
  storeId: string,
  productId: string,
  signal?: AbortSignal,
): Promise<DemandSeries> {
  const parameters = new URLSearchParams({ store_id: storeId, product_id: productId });
  return getJson<DemandSeries>(`/api/v1/demand/series?${parameters.toString()}`, signal);
}

export function getDemandSignals(
  storeId: string,
  productId: string,
  signal?: AbortSignal,
): Promise<DemandSignal[]> {
  const parameters = new URLSearchParams({ store_id: storeId, product_id: productId });
  return getJson<DemandSignal[]>(`/api/v1/signals?${parameters.toString()}`, signal);
}

export function getInventoryPosition(
  storeId: string,
  productId: string,
  signal?: AbortSignal,
): Promise<InventoryPosition> {
  const parameters = new URLSearchParams({ store_id: storeId, product_id: productId });
  return getJson<InventoryPosition>(`/api/v1/inventory/positions?${parameters.toString()}`, signal);
}

export function getInventoryRisk(
  storeId: string,
  productId: string,
  signal?: AbortSignal,
): Promise<InventoryRisk> {
  const parameters = new URLSearchParams({ store_id: storeId, product_id: productId });
  return getJson<InventoryRisk>(`/api/v1/inventory/risks?${parameters.toString()}`, signal);
}

export function getPlanningExceptions(
  storeId: string,
  productId: string,
  signal?: AbortSignal,
): Promise<PlanningException[]> {
  const parameters = new URLSearchParams({ store_id: storeId, product_id: productId });
  return getJson<PlanningException[]>(`/api/v1/exceptions?${parameters.toString()}`, signal);
}

export function createForecastPreview(request: ForecastPreviewRequest): Promise<ForecastPreview> {
  return postJson<ForecastPreviewRequest, ForecastPreview>('/api/v1/forecast-previews', request);
}

File: backend/app/schemas/planning.py
"""HTTP schemas for catalog and demand-planning reads."""

from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class StoreResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    store_id: str
    store_code: str
    store_name: str
    city: str
    region: str
    country: str
    timezone: str


class ProductResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    product_id: str
    sku: str
    product_name: str
    category: str
    subcategory: str
    brand: str
    unit_cost: Decimal
    unit_price: Decimal


class PlanningPolicyResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    lead_time_days: int
    safety_stock: int
    reorder_point: int
    minimum_order_quantity: int
    order_multiple: int
    service_level_target: float


class DemandPointResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    period_start: date
    demand_units: int


class DemandSummaryResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    period_start: date
    period_end: date
    total_units: int
    average_monthly_units: float
    minimum_monthly_units: int
    maximum_monthly_units: int
    zero_demand_months: int


class DemandSeriesResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    store_id: str
    product: ProductResponse
    policy: PlanningPolicyResponse
    observations: list[DemandPointResponse]
    summary: DemandSummaryResponse


class ForecastPreviewRequest(BaseModel):
    model_config = ConfigDict(frozen=True)

    store_id: str = Field(min_length=1)
    product_id: str = Field(min_length=1)
    horizon_months: int = Field(default=6, ge=1, le=12)
    interval_level: int = Field(default=90, ge=50, le=99)


class ForecastMetricsResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    wape: float | None
    mase: float | None
    rmse: float
    bias: float | None
    interval_coverage: float | None
    validation_points: int


class ModelEvaluationResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    model_name: str
    status: Literal["succeeded", "failed"]
    validation_origins: int
    metrics: ForecastMetricsResponse | None
    failure_detail: str | None


class ForecastPointResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    period_start: date
    forecast_value: float
    lower_bound: float
    upper_bound: float
    horizon: int


class ForecastPreviewResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    forecast_id: str
    store_id: str
    product: ProductResponse
    selected_model: str
    training_cutoff: date
    horizon_months: int
    interval_level: int
    intermittent_demand: bool
    validation_origins: int
    selected_metrics: ForecastMetricsResponse
    evaluations: list[ModelEvaluationResponse]
    forecasts: list[ForecastPointResponse]


class DemandSignalResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    signal_id: str
    store_id: str
    product_id: str
    sku: str
    signal_type: Literal["trend", "volatility", "seasonality", "anomaly"]
    direction: Literal["growing", "declining", "stable", "volatile", "seasonal", "spike", "drop"]
    severity: Literal["info", "watch", "warning"]
    period_start: date | None
    title: str
    description: str
    metric_name: str
    metric_value: float
    baseline_value: float | None
    threshold_value: float
    evidence: dict[str, str | int | float]


class InventoryPositionResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    store_id: str
    product_id: str
    sku: str
    snapshot_at: datetime
    on_hand: int
    allocated: int
    on_order: int
    on_order_due_within_lead_time: int
    next_expected_receipt_at: datetime | None
    lead_time_days: int
    safety_stock: int


class InventoryRiskResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    risk_id: str
    store_id: str
    product_id: str
    sku: str
    classification: Literal["stockout", "below_safety_stock", "healthy", "excess"]
    severity: Literal["info", "watch", "warning"]
    snapshot_at: datetime
    replenishment_arrival_at: datetime
    available_inventory: int
    on_order_due_within_lead_time: int
    forecast_demand_during_lead_time: float
    projected_inventory: float
    safety_stock: int
    coverage_months: float | None
    excess_coverage_threshold_months: float
    selected_model: str
    forecast_id: str
    training_cutoff: date
    forecast_horizon_months: int


class PlanningExceptionResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    exception_id: str
    store_id: str
    product_id: str
    sku: str
    exception_type: Literal[
        "potential_stockout",
        "below_safety_stock",
        "excess_inventory",
        "demand_spike",
        "demand_decline",
        "forecast_uncertainty",
        "forecast_bias",
    ]
    severity: Literal["medium", "high", "critical"]
    priority_score: int
    status: Literal["open", "acknowledged", "resolved", "dismissed"]
    relevant_period: date
    title: str
    description: str
    metric_name: str
    metric_value: float
    threshold_value: float
    evidence: dict[str, str | int | float]
    related_signal_id: str | None
    related_risk_id: str | None
    related_forecast_id: str | None
    created_at: datetime
